<p align="right"><a href="./C2_W2_Derivatives.ipynb">English</a> | <strong>中文</strong></p>

# Optional Lab - 导数（Derivatives）
本实验会让你对导数有更直观的理解。它会展示一种用算术方式计算导数的简单方法，还会介绍一个方便的 Python 库，让你能符号化地（symbolically）计算导数。

In [1]:
from sympy import symbols, diff

## 导数的非正式定义

导数的正式定义带着极限和 '趋于零' 的值，可能有点吓人。但其思想其实简单得多。

一个函数的导数，描述的是：当输入变量有一个小变化时，函数的输出如何变化。

我们以代价函数 $J(w)$ 为例。代价 $J$ 是输出，$w$ 是输入变量。
我们给 '小变化' 起个名字 *epsilon* 或 $\epsilon$。用这些希腊字母是因为数学里传统上用 *epsilon*($\epsilon$) 或 *delta* ($\Delta$) 表示一个小值。你可以把它想成 0.001 或别的什么小值。

$$
\begin{equation}
\text{if } w \uparrow \epsilon \text{ causes }J(w) \uparrow \text{by }k \times \epsilon \text{ then}  \\
\frac{\partial J(w)}{\partial w} = k \tag{1}
\end{equation}
$$

这只是说：如果你把函数 $J(w)$ 的输入改变一点点，而输出变化了那一点点的 $k$ 倍，那么 $J(w)$ 的导数就等于 $k$。

我们来试试。看看函数 $J(w) = w^2$ 在 $w=3$、$\epsilon = 0.001$ 处的导数

In [2]:
J = (3)**2
J_epsilon = (3 + 0.001)**2
k = (J_epsilon - J)/0.001    # difference divided by epsilon
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k:0.6f} ")

J = 9, J_epsilon = 9.006001, dJ_dw ~= k = 6.001000 


我们把输入值增加了一点点（0.001），使输出从 9 变成 9.006001，增量是输入增量的 6 倍。参照上面 (1)，这说明 $k=6$，所以 $\frac{\partial J(w)}{\partial w} \approx 6$。如果你熟悉微积分，就知道用符号写作 $\frac{\partial J(w)}{\partial w} = 2 w$，在 $w=3$ 时即为 6。我们上面的计算不恰好是 6，因为要完全正确，$\epsilon$ 需要[无穷小](https://www.dictionary.com/browse/infinitesimally)、或者说非常非常小。这就是我们用 $\approx$ 或 ~= 而非 = 的原因。我们看看把 $\epsilon$ 变小会怎样。

In [3]:
J = (3)**2
J_epsilon = (3 + 0.000000001)**2
k = (J_epsilon - J)/0.000000001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 9, J_epsilon = 9.000000006, dJ_dw ~= k = 6.000000496442226 


随着我们减小 $\epsilon$ 的大小，这个值接近恰好为 6。你可以试着进一步减小它。

## 求符号导数
在反向传播中，知道简单函数在任意输入值处的导数很有用。换句话说，我们想知道 '符号（symbolic）' 导数、而非 '算术（arithmetic）' 导数。符号导数的一个例子是 $\frac{\partial J(w)}{\partial w} = 2 w$，即上面 $J(w) = w^2$ 的导数。有了符号导数，你就能求出任意输入值 $w$ 处导数的值。

如果你上过微积分课，就熟悉数学家发展出的许多[求导法则](https://en.wikipedia.org/wiki/Differentiation_rules#Power_laws,_polynomials,_quotients,_and_reciprocals)，用来对一个表达式求导。事实上，这个过程已被符号微分程序自动化了。Python 里的一个例子是 [SymPy](https://www.sympy.org/en/index.html) 库。我们看看怎么用它。

### $J = w^2$
定义 python 变量及其符号名。

In [4]:
J, w = symbols('J, w')

定义并打印表达式。注意 SymPy 会生成一个 [latex](https://en.wikibooks.org/wiki/LaTeX/Mathematics) 字符串，从而渲染出漂亮易读的方程。

In [5]:
J=w**2
J

w**2

用 SymPy 的 `diff` 对 $J$ 关于 $w$ 求导。注意结果与我们前面的例子一致。

In [6]:
dJ_dw = diff(J,w)
dJ_dw

2*w

通过把符号值 '替换（substitute）' 成数值，在几个点上求导数的值。第一个例子里，$w$ 被替换成 $2$。

In [7]:
dJ_dw.subs([(w,2)])    # derivative at the point w = 2

4

In [8]:
dJ_dw.subs([(w,3)])    # derivative at the point w = 3

6

In [9]:
dJ_dw.subs([(w,-3)])    # derivative at the point w = -3

-6

## $J = 2w$

In [10]:
w, J = symbols('w, J')

In [11]:
J = 2 * w
J

2*w

In [12]:
dJ_dw = diff(J,w)
dJ_dw

2

In [13]:
dJ_dw.subs([(w,-3)])    # derivative at the point w = -3

2

把它与算术计算作比较

In [14]:
J = 2*3
J_epsilon = 2*(3 + 0.001)
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 6, J_epsilon = 6.002, dJ_dw ~= k = 1.9999999999997797 


对函数 $J=2w$，很容易看出：无论 $w$ 的起始值是多少，$w$ 的任何变化都会使输出 $J$ 变化其 2 倍。我们的 NumPy 和算术结果证实了这一点。

## $J = w^3$

In [15]:
J, w = symbols('J, w')

In [16]:
J=w**3
J

w**3

In [17]:
dJ_dw = diff(J,w)
dJ_dw

3*w**2

In [18]:
dJ_dw.subs([(w,2)])   # derivative at the point w=2

12

把它与算术计算作比较

In [19]:
J = (2)**3
J_epsilon = (2+0.001)**3
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 8, J_epsilon = 8.012006000999998, dJ_dw ~= k = 12.006000999997823 


## $J = \frac{1}{w}$

In [20]:
J, w = symbols('J, w')

In [21]:
J= 1/w
J

1/w

In [22]:
dJ_dw = diff(J,w)
dJ_dw

-1/w**2

In [23]:
dJ_dw.subs([(w,2)])

-1/4

把它与算术计算作比较

In [24]:
J = 1/2
J_epsilon = 1/(2+0.001)
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 0.5, J_epsilon = 0.49975012493753124, dJ_dw ~= k = -0.2498750624687629 


## $J = \frac{1}{w^2}$

In [25]:
J, w = symbols('J, w')

如果有时间，试着对函数 $J = \frac{1}{w^2}$ 重复上面的步骤，并在 w=4 处求值

In [29]:
J= 1/w**2
dJ_dw = diff(J,w)
dJ_dw


-2/w**3

In [30]:
dJ_dw.subs([(w,4)])

-1/32

把它与算术计算作比较

In [26]:
J = 1/4**2
J_epsilon = 1/(4+0.001)**2
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 0.0625, J_epsilon = 0.06246876171484496, dJ_dw ~= k = -0.031238285155041345 


<details>
  <summary><font size="3" color="darkgreen"><b>点击查看提示</b></font></summary>

```python
J= 1/w**2
dJ_dw = diff(J,w)
dJ_dw.subs([(w,4)])
```


</details>

## 恭喜！
如果你走完了上面的例子，你就理解了：导数描述的是"函数输入的一个小变化所导致的函数输出的变化"。你也会用 python 里的 *SymPy* 来求函数的符号导数了。